src/libraries/System.Private.CoreLib/src/System/SpanHelpers.T.cs

```diff
@@ -3846,7 +3847,23 @@ internal static int LastIndexOfAnyExceptInRangeUnsignedNumber<T>(ref T searchSpa
                     ref T oneVectorAwayFromEnd = ref Unsafe.Subtract(ref end, Vector128<T>.Count);
                     do
                     {
-                        count += BitOperations.PopCount(Vector128.Equals(Vector128.LoadUnsafe(ref current), targetVector).ExtractMostSignificantBits());
+                        if (AdvSimd.Arm64.IsSupported)
+                        {
+                            Vector128<T> mask = Vector128.Equals(Vector128.LoadUnsafe(ref current), targetVector);
+                            ulong matches = AdvSimd.ShiftRightLogicalNarrowingLower(mask.AsUInt16(), 4).AsUInt64().ToScalar();
+                            count += BitOperations.PopCount(matches) >> (Unsafe.SizeOf<T>() switch
+                            {
+                                1 => 2,
+                                2 => 3,
+                                4 => 4,
+                                _ => 5
```

Comment:
I could not come up with better way to write this.

Reply:
```c#
count += (int)((uint)BitOperations.PopCount(matches) / (4 * sizeof(T)));
```

Reply:
Thanks, it seems JIT can't completely fold this:
```cs
[MethodImpl(MethodImplOptions.AggressiveOptimization | MethodImplOptions.NoInlining)]
static int CountAdvSimd<T>(ReadOnlySpan<T> span, T value)
{
    var eqmask = Vector128.Equals(Vector128.Create(span), Vector128.Create(value));
    var matches = AdvSimd.ShiftRightLogicalNarrowingLower(eqmask.AsUInt16(), 4).AsUInt64().ToScalar();
    // or (int)((uint)BitOperations.PopCount(matches) / (4 * sizeof(T)));
    return BitOperations.PopCount(matches) >> (Unsafe.SizeOf<T>() switch
    {
        1 => 2,
        2 => 3,
        4 => 4,
        _ => 5
    });
}
```

Diff where T is `ushort`: https://www.diffchecker.com/hzMEBfhM/

Reply:
I don't see any codegen differences locally

Reply:
The version by @MihaZupan is faster 🤯
diff:
```diff
G_M000_IG04:                ;; offset=0028H
            ldr     q17, [x0]
            cmeq    v17.16b, v17.16b, v16.16b
            shrn    v17.8b, v17.8h, #4
            umov    x5, v17.d[0]
            mov     v17.d[0], x5
            cnt     v17.8b, v17.8b
            addv    b17, v17.8b
            umov    w5, v17.b[0]
-           add     w3, w3, w5,  LSR #2
+           mov     w5, w5
+           lsr     x5, x5, #2
+           add     w3, w5, w3
            add     x0, x0, #16
            cmp     x0, x2
            bls     G_M000_IG04
```

I think I know why. According to https://dougallj.github.io/applecpu/firestorm-int.html, `ADD (register, lsr, 32-bit)` has reciprocal TP of 0.333 per cycle with 2 cycle latency. However, both `ADD (register, 32-bit)` and `LSR (immediate, 32-bit)` are 0.111 of reciprocal TP with 1 cycle lat. This looks like a stall on a backwards jump because *if I forcibly use V256 (V128x2) version it is marginally faster too despite codegen*. 

According to napkin math, slower version handles around 5.8~6 bytes per cycle while the faster one is ~8. If the above assumption is correct, it would explain the difference.

All in all this was very counterintuitive, TIL.